[ao][fx]refactor memoryless from user facing APIs

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #73560

Summary: removed memoryless from user facing APIs and replaced with logic through the use of compute_dtype.
if compute_dtype isn't specified there are no changes to how this works, if compute_dtype is specified
then its used as the observer dtype (rather than the dtype which is treated as the activation_dtype).
Finally if compute_dtype is a quantized dtype and dtype=float (which is the designation for dynamic quantization we
came up with) then a memoryless observer is created.

Test Plan: python test/test_quantization.py

Differential Revision: [D34546005](https://our.internmc.facebook.com/intern/diff/D34546005)

HDCharles committed:
[ao][fx]refactor memoryless from user facing APIs

removed memoryless from user facing APIs and replaced with logic through the use of compute_dtype.
if compute_dtype isn't specified there are no changes to how this works, if compute_dtype is specified
then its used as the observer dtype (rather than the dtype which is treated as the activation_dtype).
Finally if compute_dtype is a quantized dtype and dtype=float (which is the designation for dynamic quantization we
came up with) then a memoryless observer is created.

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][fx]refactor memoryless from user facing APIs"


removed memoryless from user facing APIs and replaced with logic through the use of compute_dtype.
if compute_dtype isn't specified there are no changes to how this works, if compute_dtype is specified
then its used as the observer dtype (rather than the dtype which is treated as the activation_dtype).
Finally if compute_dtype is a quantized dtype and dtype=float (which is the designation for dynamic quantization we
came up with) then a memoryless observer is created.

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][fx]refactor memoryless from user facing APIs"


removed memoryless from user facing APIs and replaced with logic through the use of compute_dtype.
if compute_dtype isn't specified there are no changes to how this works, if compute_dtype is specified
then its used as the observer dtype (rather than the dtype which is treated as the activation_dtype).
Finally if compute_dtype is a quantized dtype and dtype=float (which is the designation for dynamic quantization we
came up with) then a memoryless observer is created.

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][fx]refactor memoryless from user facing APIs"


Summary: removed memoryless from user facing APIs and replaced with logic through the use of compute_dtype.
if compute_dtype isn't specified there are no changes to how this works, if compute_dtype is specified
then its used as the observer dtype (rather than the dtype which is treated as the activation_dtype).
Finally if compute_dtype is a quantized dtype and dtype=float (which is the designation for dynamic quantization we
came up with) then a memoryless observer is created.

Test Plan: python test/test_quantization.py

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34546005).

jerryzh168 commented on torch/ao/quantization/observer.py:
```diff
@@ -346,8 +420,8 @@ class MinMaxObserver(_ObserverBase):
         quant_max: Maximum quantization value. If unspecified, it will follow the 8-bit setup.
-        memoryless: Boolean that controls whether observer removes old data when a new input is seen.
-                    This is most useful for simulating dynamic quantization, especially during QAT.
+        compute_dtype: if compute_dtype is not set, dtype is the compute_dtype, if compute_dtype
```
I think a better way to describe this is: if compute_dtype is None, then both activation_dtype and compute_dtype are dtype, if compute_dtype is not None, then we may have a different compute_dtype than the activation_dtype

jerryzh168 commented on torch/ao/quantization/observer.py:
```diff
@@ -406,68 +478,26 @@ def __init__(
+                                       torch.qint32,) and dtype == torch.float32
         super(MinMaxObserver, self).__init__(
-            dtype=dtype,
+            dtype=dtype if compute_dtype is None else compute_dtype,
```
maybe we can always pass in compute_dtype, but set compute_dtype to dtype when compute_dtype is None

HDCharles replied:
ya, I made this change, it still feels a little weird to set compute_dtype to dtype and then set dtype to compute_dtype when we pass it in, but I'm not sure if there's a better way.

jerryzh168 reviewed (commented):
Should we change the default_qconfig for dynamic quantization as well? https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/observer.py#L1451
This might be a BC-breaking change

vkuzo commented:
Just curious, is the change in inheritance related to the goal of this diff, or is it just a refactor to reduce duplicated code?

1. If the inheritance stuff is not related, could we split that into a separate PR?  Combining it makes this hard to read, which 
is especially important to avoid when changing user APIs.
2. Should we consider keeping the inheritance as is?  Seems like the inheritance change improves code reuse but decreases readability, we already have some non-trivial inheritance for observers and this makes it even more complex.  There are other strategies that could be used to improve code reuse without creating new inheritance patterns.

vkuzo commented on test/quantization/core/test_workflow_module.py:
```diff
@@ -400,7 +400,7 @@ def test_zero_numel(self):
 
     def _test_memoryless(self, obs_class):
-        obs = obs_class(memoryless=True)
+        obs = obs_class(dtype=torch.float32, compute_dtype=obs_class().dtype)
```
if we are changing the user facing API, can we also rename the test case?

HDCharles replied:
it still creates a memoryless observer, its just hidden from the API

vkuzo reviewed (changes requested):
the "memoryless" part looks good to me, could we follow up on the inheritance question?

HDCharles committed:
Update on "[ao][fx]refactor memoryless from user facing APIs"


Summary: removed memoryless from user facing APIs and replaced with logic through the use of compute_dtype.
if compute_dtype isn't specified there are no changes to how this works, if compute_dtype is specified
then its used as the observer dtype (rather than the dtype which is treated as the activation_dtype).
Finally if compute_dtype is a quantized dtype and dtype=float (which is the designation for dynamic quantization we
came up with) then a memoryless observer is created.

Test Plan: python test/test_quantization.py

Differential Revision: [D34546005](https://our.internmc.facebook.com/intern/diff/D34546005)

[ghstack-poisoned]

HDCharles commented:
@vkuzo 
> If the inheritance stuff is not related, could we split that into a separate PR? Combining it makes this hard to read, which
is especially important to avoid when changing user APIs.
Should we consider keeping the inheritance as is? Seems like the inheritance change improves code reuse but decreases readability, we already have some non-trivial inheritance for observers and this makes it even more complex. There are other strategies that could be used to improve code reuse without creating new inheritance patterns.

The only other way I see of doing this without the inheritance seemed more confusing. i.e. we take dtype and compute_dtype as arguments and then overwrite dtype with compute_dtype if its specified. Also set memoryless based on the two args. That seemed worse when I originally wrote this but I can make that change if you think its an improvement. This way seemed easier since the old public class is excatly the same as the new private class.

@jerryzh168 
> Should we change the default_qconfig for dynamic quantization as well? https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/observer.py#L1451 This might be a BC-breaking change

it works without the change, the placeholder observer doesn't have a memoryless argument, should it be changed to be able to handle the same compute_dtype logic as the other observers?

HDCharles commented:
see https://github.com/pytorch/pytorch/pull/73947
